File: ios/RCTNendVideoNativeAd.m
//
//  RCTNendVideoNativeAd.m
//  AwesomeProject
//

#import "RCTNendVideoNativeAd.h"
#import <React/RCTUIManager.h>

static NSString * const kRCTNendVideoNativeAdClickOptionFullScreen = @"FullScreen";
static NSString * const kRCTNendVideoNativeAdClickOptionLP = @"LP";
static NSString * const kRCTNendVideoNativeAdEventName = @"VideoNativeAdEventListener";

@import NendAd;

@interface NADNativeVideo (RCTExtension)

@end

@implementation NADNativeVideo (RCTExtension)

- (NSDictionary<NSString *, id> *)getJSONWithReferenceId:(NSInteger)refId
{
  return @{@"logoImageUrl": self.logoImageUrl,
           @"title": self.title,
           @"description": self.explanation,
           @"advertiserName": self.advertiserName,
           @"userRating": @(self.userRating),
           @"userRatingCount": @(self.userRatingCount),
           @"callToAction": self.callToAction,
           @"referenceId": @(refId)};
}

@end

#pragma mark -

@interface RCTNendVideoNativeAd () <NADNativeVideoDelegate>

@property (nonatomic, strong) NSMutableDictionary<NSString *, NADNativeVideoLoader *> *loaderCache;
@property (nonatomic, strong) NSMutableDictionary<NSNumber *, NADNativeVideo *> *videoAdCache;
@property (nonatomic, strong) NSMapTable<NADNativeVideo *, NSNumber *> *videoAdToReferenceId;
@property (nonatomic) NSInteger referenceId;
@property (nonatomic) BOOL hasListeners;

@end

@implementation RCTNendVideoNativeAd

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}

RCT_EXPORT_MODULE(NendVideoNativeAd)

RCT_EXPORT_METHOD(initialize:(NSString *)spotId apiKey:(NSString *)apiKey clickOption:(NSInteger)clickOption)
{
  NADNativeVideoLoader *loader = self.loaderCache[spotId];
  if (!loader) {
    loader = [[NADNativeVideoLoader alloc] initWithSpotId:spotId
                                                   apiKey:apiKey
                                              clickAction:(NADNativeVideoClickAction)clickOption];
    self.loaderCache[spotId] = loader;
  }
}

RCT_EXPORT_METHOD(loadAd:(NSString *)spotId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  NADNativeVideoLoader *loader = self.loaderCache[spotId];
  if (loader) {
    __weak typeof(self) weakSelf = self;
    [loader loadAdWithCompletionHandler:^(NADNativeVideo * _Nullable ad, NSError * _Nullable error) {
      if (ad) {
        ad.delegate = weakSelf;
        NSInteger refId = ++weakSelf.referenceId;
        weakSelf.videoAdCache[@(refId)] = ad;
        [weakSelf.videoAdToReferenceId setObject:@(refId) forKey:ad];
        resolve([ad getJSONWithReferenceId:refId]);
      } else {
        reject(@"", @"", error);
      }
    }];
  }
}

RCT_EXPORT_METHOD(registerClickableViews:(nonnull NSNumber *)refernceId tags:(NSArray<NSNumber *> *)tags)
{
  NADNativeVideo *videoAd = self.videoAdCache[refernceId];
  if (videoAd && tags && tags.count > 0) {
    NSMutableArray<UIView *> *views = [NSMutableArray array];
    [tags enumerateObjectsUsingBlock:^(NSNumber * _Nonnull tag, NSUInteger idx, BOOL * _Nonnull stop) {
      UIView *view = [self.bridge.uiManager viewForReactTag:tag];
      if (view) {
        [views addObject:view];
      }
    }];
    if (views.count > 0) {
      [videoAd registerInteractionViews:views];
    }
  }
}

RCT_EXPORT_METHOD(destroyLoader:(NSString *)spotId)
{
  [self.loaderCache removeObjectForKey:spotId];
}

RCT_EXPORT_METHOD(destroyAd:(nonnull NSNumber *)refernceId)
{
  [self.videoAdCache removeObjectForKey:refernceId];
}

- (NSDictionary *)constantsToExport
{
  return @{kRCTNendVideoNativeAdClickOptionFullScreen: @1, kRCTNendVideoNativeAdClickOptionLP: @2};
}

- (NSArray<NSString *> *)supportedEvents
{
  return @[kRCTNendVideoNativeAdEventName];
}

- (void)startObserving
{
  self.hasListeners = YES;
}

- (void)stopObserving
{
  self.hasListeners = NO;
}

- (instancetype)init
{
  self = [super init];
  if (self) {
    _loaderCache = [NSMutableDictionary new];
    _videoAdCache = [NSMutableDictionary new];
    _videoAdToReferenceId = [NSMapTable weakToStrongObjectsMapTable];
    _referenceId = 0;
    _hasListeners = NO;
  }
  return self;
}

- (NADNativeVideo *)getVideoAdWithRefernceId:(NSInteger)referenceId
{
  return self.videoAdCache[@(referenceId)];
}

- (void)sendEventWithVideoAd:(NADNativeVideo *)videoAd eventType:(NSString *)eventType
{
  NSNumber *refId = [self.videoAdToReferenceId objectForKey:videoAd];
  if (refId && self.hasListeners) {
    [self sendEventWithName:kRCTNendVideoNativeAdEventName body:@{@"refId": refId, @"eventType": eventType}];
  }
}

#pragma mark - NADNativeVideoDelegate

- (void)nadNativeVideoDidImpression:(NADNativeVideo * _Nonnull)ad
{
  [self sendEventWithVideoAd:ad eventType:@"onImpression"];
}

- (void)nadNativeVideoDidClickAd:(NADNativeVideo * _Nonnull)ad
{
  [self sendEventWithVideoAd:ad eventType:@"onClickAd"];
}

- (void)nadNativeVideoDidClickInformation:(NADNativeVideo * _Nonnull)ad
{
  [self sendEventWithVideoAd:ad eventType:@"onClickInformation"];
}

@end

File: ios/RNNendBridger/RCTNendVideoNativeAd.swift
//
//  RCTNendVideoNativeAd.swift
//  AwesomeProject
//

import UIKit
import NendAd

@objc(RCTNendVideoNativeAd)
class RCTNendVideoNativeAd: RCTEventEmitter {
  private let kRCTNendVideoNativeAdEventName = "VideoNativeAdEventListener"
  private var loaderCache = [String: NADNativeVideoLoader]()
  private var videoAdCache = [Int: NADNativeVideo]()
  private let videoAdToReferenceId = NSMapTable<NADNativeVideo,  NSNumber>(keyOptions: .weakMemory, valueOptions: .strongMemory)
  private var referenceId = 0
  private var hasListeners = false
  
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  override var methodQueue: DispatchQueue! {
    return DispatchQueue.main
  }
  
  override func constantsToExport() -> [AnyHashable : Any]! {
    return ["FullScreen": 1, "LP": 2]
  }
  
  override func supportedEvents() -> [String]! {
    return [kRCTNendVideoNativeAdEventName]
  }
  
  override func startObserving() {
    hasListeners = true
  }
  
  override func stopObserving() {
    hasListeners = false
  }
  
  @objc(initialize:apiKey:clickOption:)
  func initialize(spotId: String, apiKey: String, clickOption: Int) {
    if loaderCache[spotId] == nil, let option = NADNativeVideoClickAction(rawValue: clickOption) {
      let loader = NADNativeVideoLoader(spotId: spotId, apiKey: apiKey, clickAction: option)
      loaderCache[spotId] = loader
    }
  }
  
  @objc(loadAd:resolver:rejecter:)
  func loadAd(spotId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let loader = loaderCache[spotId] else { return }
    loader.loadAd { [weak self] (ad, error) in
      guard let `self` = self else { return }
      if let videoAd = ad {
        videoAd.delegate = self
        self.referenceId += 1
        self.videoAdCache[self.referenceId] = videoAd
        self.videoAdToReferenceId.setObject(NSNumber(integerLiteral: self.referenceId), forKey: videoAd)
        resolver(videoAd.toJSON(refId: self.referenceId))
      } else if let e = error {
        rejecter("", "", e)
      }
    }
  }
  
  @objc(registerClickableViews:reactTags:)
  func registerClickableViews(refernceId: Int, reactTags: [NSNumber]) {
    if let videoAd = videoAdCache[refernceId] {
      videoAd.registerInteractionViews(reactTags.flatMap { bridge.uiManager.view(forReactTag: $0) })
    }
  }
  
  @objc(setUserId:userId:)
  func setUserId(spotId: String, userId: String) {
    if let loader = loaderCache[spotId] {
      loader.userId = userId
    }
  }

  @objc(setUserFeature:refId:)
  func setUserFeature(spotId: String, refId: Int) {
    if let module = bridge.module(for: RCTNendUserFeature.self) as? RCTNendUserFeature,
      let loader = loaderCache[spotId],
      let feature = module.getUserFeature(refId: refId) {
      loader.userFeature = feature
    }
  }
  
  @objc(destroyLoader:)
  func destroyLoader(spotId: String) {
    loaderCache.removeValue(forKey: spotId)
  }
  
  @objc(destroyAd:)
  func destroyAd(refernceId: Int) {
    videoAdCache.removeValue(forKey: refernceId)
  }
  
  fileprivate func sendEvent(videoAd: NADNativeVideo, eventType: String) {
    if let refId = videoAdToReferenceId.object(forKey: videoAd)?.intValue, hasListeners {
      sendEvent(withName: kRCTNendVideoNativeAdEventName, body: ["refId": refId, "eventType": eventType])
    }
  }
  
  func getVideoAd(refernceId: Int) -> NADNativeVideo? {
    return videoAdCache[refernceId]
  }
}

extension RCTNendVideoNativeAd: NADNativeVideoDelegate {
  func nadNativeVideoDidImpression(_ ad: NADNativeVideo) {
    sendEvent(videoAd: ad, eventType: "onImpression")
  }
  
  func nadNativeVideoDidClickAd(_ ad: NADNativeVideo) {
    sendEvent(videoAd: ad, eventType: "onClickAd")
  }
  
  func nadNativeVideoDidClickInformation(_ ad: NADNativeVideo) {
    sendEvent(videoAd: ad, eventType: "onClickInformation")
  }
}

extension NADNativeVideo {
  fileprivate func toJSON(refId: Int) -> [String: Any] {
    return ["logoImageUrl": logoImageUrl!,
            "title": title!,
            "description": explanation!,
            "advertiserName": advertiserName!,
            "userRating": userRating,
            "userRatingCount": userRatingCount,
            "callToAction": callToAction!,
            "referenceId": refId]
  }
}
